### ISSUE TITLE
YOLOv3模块中PascalVOC生成器的原理问题

### ISSUE THREAD

**Issue by KingPengcj** [xuannianz/FSAF#12]
亲爱的博主，您好，想咨询一下，这句话中PascalVOC生成器的原理是什么
train_generator = PascalVocGenerator(
            args.pascal_path,
            'trainval',
            skip_difficult=True,
            misc_effect=misc_effect,
            visual_effect=visual_effect,
            anchors_path='voc_anchors_416.txt',
            multi_scale=args.multi_scale,
            **common_args
        )

**Comment by xuannianz** [comment]
每次迭代都会调用 generator 的 \__getitem__ 方法， \__getitem__ 返回网络所需要的输入和输出。

**Comment by KingPengcj** [comment]
如果换成自己数据集的话，应该对标注好的数据集做什么样的处理，作为网络的的输入

**Comment by xuannianz** [comment]
根据你的 annotations 保存的格式 (coco, pascal, csv)， 然后跑 train.py 会指定就行。
三者分别对应三个类 PascalGenerator, CSVGenerator, CocoGenerator, 这三个类都会继承 Geneartor 这个类，具体为网络准备输入输出就在这个类中完成的，不需要再做什么操作了，训练之前你可以检查一下生成的输入输出对不对。

**Comment by KingPengcj** [comment]
十分感谢您的指导，想再问一下，yolo_body函数中的FSAFTarget（）函数是什么作用，这里有点没看懂。
batch_cls_target, batch_cls_mask, batch_cls_num_pos, batch_regr_target, batch_regr_mask = FSAFTarget(
        num_classes=num_classes,
        name='fsaf_target')(
        [batch_gt_box_levels, grid_shapes_input, gt_boxes_input])

**Comment by xuannianz** [comment]
这个就是生成的网络的真值，后面用来计算 loss

**Comment by KingPengcj** [comment]
好的，谢谢您细心的回复，有一个不解困扰鄙人许久，对于YOLOv3原版本，已标注数据所转换的真值y_true的结构如下：
[Tensor("input_2:0", shape=(?, 13, 13, 3, 5+num_class), dtype=float32),
  Tensor("input_3:0", shape=(?, 26, 26, 3, 5+num_class), dtype=float32),
  Tensor("input_4:0", shape=(?, 52, 52, 3, 5+num_class), dtype=float32)]
那么anchor-free模块的y_true的结构是什么样子的，跪求博主指点一二。

**Comment by xuannianz** [comment]
> 十分感谢您的指导，想再问一下，yolo_body函数中的FSAFTarget（）函数是什么作用，这里有点没看懂。
> batch_cls_target, batch_cls_mask, batch_cls_num_pos, batch_regr_target, batch_regr_mask = FSAFTarget(
> num_classes=num_classes,
> name='fsaf_target')(
> [batch_gt_box_levels, grid_shapes_input, gt_boxes_input])

batch_cls_target 和 batch_regr_target 就是 y_true, 一个是目标分类，一个是目标框回归

**Comment by KingPengcj** [comment]
十分感谢您的回答，太谢谢您了，博主可以讲讲yolov3中anchor-free+anchor-based的数据生成器该融合的方法吗

**Comment by xuannianz** [comment]
没太懂，数据生成器依赖你网络的设计，返回的结果和网络输入输出对应起来就行。

**Comment by KingPengcj** [comment]
博主在代码里面设置的bach_size=1，想问一下，如果bach_size>1的情况下，下面的代码如何修改，可以提供一个思路吗？
gt_boxes_input = Input(shape=(config.MAX_NUM_GT_BOXES, 5), name='gt_boxes_input')
grid_shapes_input = Input((len(config.STRIDES), 2), dtype='int32', name='grid_shapes_input')
batch_gt_box_levels = LevelSelect(name='level_select')(
        [batch_cls_pred, batch_regr_pred, grid_shapes_input, gt_boxes_input]) batch_cls_target, batch_cls_mask, batch_cls_num_pos, batch_regr_target, batch_regr_mask = FSAFTarget(
        num_classes=num_classes,name='fsaf_target')(
        [batch_gt_box_levels, grid_shapes_input, gt_boxes_input])
    focal_loss_graph = focal_with_mask()
    iou_loss_graph = iou_with_mask()
    cls_loss = Lambda(focal_loss_graph,output_shape=(1,),name="cls_loss")(
        [batch_cls_target, batch_cls_pred, batch_cls_mask, batch_cls_num_pos])
    regr_loss = Lambda(iou_loss_graph,output_shape=(1,), name="regr_loss")
       ([batch_regr_target, batch_regr_pred, batch_regr_mask])
    model = Model(inputs=[image_input, gt_boxes_input, grid_shapes_input],
                  outputs=[cls_loss, regr_loss],name='fsaf')

**Comment by KingPengcj** [comment]
再次谢谢您的帮助，万分感谢
